### DISCUSSION TITLE
Adding your own custom transforms

### DISCUSSION THREAD

**Discussion by vikashg** [Project-MONAI/MONAI#3592] [General]
Hello Team, 
I was wondering if there is resource guide to adding custom transforms ? 
If we need to add new transforms should we be looking at the existing transforms and creating a new one and adding the appropriate hooks at different places ? or is there any better method to do that. 

Thanks 
Vikash

**Comment by Nic-Ma** [comment]
Hi @vikashg ,

Please check the "How do I create a custom transform?" part in this tutorial:
https://github.com/Project-MONAI/MONAIBootcamp2021/blob/main/day1/1.%20Getting%20Started%20with%20MONAI.ipynb

Thanks.

**Reply by vikashg** [reply]
Thanks @Nic-Ma

**Reply by vikashg** [reply]
Hi @Nic-Ma, 
I am trying to write a custom transform, which works on its own, but when I try to use it with `Compose`, I get the following error
```
  File "/raid/Vikash/Tools/MONAI/monai/transforms/compose.py", line 168, in __call__
    input_ = apply_transform(_transform, input_, self.map_items, self.unpack_items)
  File "/raid/Vikash/Tools/MONAI/monai/transforms/transform.py", line 106, in apply_transform
    raise RuntimeError(f"applying transform {transform}") from e
RuntimeError: applying transform <__main__.LoadImageList object at 0x7fe12bbe1520>
```

The name of the transform is `LoadImageList`. Any thoughts. I am not posting the whole code here because its bit long. But, let me know if you needed it I can post it here.
Thanks

**Reply by ericspod** [reply]
Can you post your code into a gist and link it here?

**Reply by vikashg** [reply]
Here is the [gist](https://gist.github.com/vikashg/9b2f8f028b8472d599c871920ab4b4dc) I have tried to give as much information in the code. But please let me know if more information is needed.

**Reply by ericspod** [reply]
How is this transform used? If your data to be passed into a Dataset object is a list of filenames then this transform will be given one filename at a time rather than whole lists, so your initial data should list of lists of filenames. In your example code:

```python
transform = Compose([LoadImageList()])
img_array = transform(listOfImages)
```

`Compose` will break the list of images into individual entries and passed those to the transform sequence then recollate the results together. Something like this should work:

```python
transform = Compose([LoadImageList()])
img_array = transform([listOfImages])  # single item dataset
```

**Reply by vikashg** [reply]
The training data will be list of list of images... I see what you did there. Yes If I call the transform as you mentioned it works. I feel stupid :) 
Thanks @ericspod. While we are on the topic, as you can see I am kind of copy-pasting bunch of code from the `LoadImage` transform. I was wondering if you have a better solution.

**Reply by ericspod** [reply]
You could inherit from `LoadImage` directly and override the methods you want to change. I think your constructor and that for `LoadImage` are the same so that should work.

**Reply by vikashg** [reply]
Ya I could do that. For some reason, I didnt want to inherit from `LoadImage`. Thanks a ton. It solves a lot of problems I was struggling with for 2 days.
